### ISSUE TITLE
Merge pdf files. Merged files are not released and cannot be deleted

### ISSUE THREAD

**Issue by lako12** [LibrePDF/OpenPDF#1112]
I have a Java method that needs to read the PDF files inside a folder and create a file corresponding to the merge of all those inside the folder. Finally I have to delete the files and leave only the result in the folder.
The problem is that once the final file is created, I can't delete all the other files used for the merge, as if it doesn't release them after use. I also tried to delete them manually from the file explorer, but I can't (as long as the application remains running). I'm using this version:
```
        <dependency>
            <groupId>com.github.librepdf</groupId>
            <artifactId>openpdf</artifactId>
            <version>2.0.1</version>
        </dependency>
```
My Code:

```
        File[] files = directory.listFiles();
        if (files != null && files.length > 0) {
            try (Document document = new Document()) {
                PdfCopy copy = new PdfCopy(document, new FileOutputStream(fileName));
                document.open();

                for (File file : files) {
                    try (PdfReader reader = new PdfReader(file.getAbsolutePath())) {
                        for (int i = 1; i <= reader.getNumberOfPages(); i++) {
                            copy.addPage(copy.getImportedPage(reader, i));
                        }
                        copy.freeReader(reader);
                    }
                }
            }
        }
```

if after this piece of code I try to delete all the files in the "files" list by calling the .delete() method, it doesn't work

I also tried removing the try wth resources and manually closing reader, document copy etc... I tried inserting the PdfCopy in the try with resources together with document, but nothing, I can't find a solution...

**Comment by asturio** [comment]
This really looks like the files are still open by the process. I assume you are also using Windows, as this OS is known for keeping resources locked.

Important is that the documents you want to delete are all closed.

You may check if the "closeStream" flag of the Objects are really "true". Maybe it is being set somewhere to "false".

Please check if the 
```java
                    currentPdfReaderInstance.getReader().close();
                    currentPdfReaderInstance.getReaderFile().close();
```
are really being executed in `freeReader` and/or `getImportedPage`. I think if they are called in freeReader, they shouldn't be called in getImportedPage

**Comment by lako12** [comment]
yes, i am usign widnows.
this methods are called:
```
  currentPdfReaderInstance.getReader().close();
currentPdfReaderInstance.getReaderFile().close();
```
but they don't do anything because they don't respect the conditions

**Comment by sa-sh** [comment]
Hello guys, I want to share my experience with PdfReader.
i noticed that issue happens only when it uses MappedRandomAccessFile (which is by default (Document.plainRandomAccess = false). 
loos like issue caused by FileChannel (used in MappedRandomAccessFile) which in java 11+ may be released/closed latter (by GC).

as workaround you can control how file opened for PdfReader
i prefer to use RandomAccessFileOrArray directly, to control how file is loaded, small files better to load into memory
```
for (File file : files) {
   boolean forceRead = file.length() <= 10_000_000; // limit memory usage, set greater value if your app can use large heap
   boolean plainRandomAccess = true; // do not use MappedRandomAccessFile, Note: this flag ignored when forceRead=true
   try (RandomAccessFileOrArray raf = new RandomAccessFileOrArray(file.getPath(), forceRead, plain);
      PdfReader reader = new PdfReader(raf, null)) {
      //  do copy
  }
}
```

**Comment by christianGen** [comment]
Here is a discussion on how to force the liberation of mapped files:

https://stackoverflow.com/questions/2972986/how-to-unmap-a-file-from-memory-mapped-using-filechannel-in-java

TL;DR: pretty much impossible for newer jvms.

**Comment by andreasrosdal** [comment]
Pull requests to improve this is welcome. @rzam

**Comment by rzam** [comment]
@andreasrosdal Since memory mapping is "bugged" at the OS level I'm not sure there *is* a way other than forgoing memory mapping itself.

Though I was checking the code and I noticed that the [`Document.plainRandomAccess`](https://github.com/LibrePDF/OpenPDF/blob/00afd24a1e44520dc929187cf3840381f5ea8160/openpdf/src/main/java/com/lowagie/text/Document.java#L117) field mentioned in this thread is a public non-final field *and* that it seems to affect memory mapping throughout the whole library (font loading included), so I guess it could be used as a workaround when working on Windows machines? I'm not sure if it was intentional since it doesn't seem to be documented.

I've done some tests and a simple `Document.plainRandomAccess = true;` before the first rendering seems to fix the issue I was experiencing when loading fonts.

**Comment by cselagea** [comment]
The [`FileChannel#map` documentation](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/nio/channels/FileChannel.html#map(java.nio.channels.FileChannel.MapMode,long,long)) states that "the buffer and the mapping that it represents will remain valid until the buffer itself is garbage-collected", which aligns with what @sa-sh said.

However, there is a more predictable way to deallocate the memory and close the file handle — call `sun.misc.Unsafe#invokeCleaner`. In fact, prior to 2.0.4 this is exactly what OpenPDF was doing. It's not clear whether this was intentional, but the buffer cleaning code was removed in #1319.

Perhaps it was because the `cleanJava11` method was broken? (It was silently failing due to `java.lang.SecurityException: Unsafe`)

I propose to reintroduce the buffer cleaning mechanism and fix it by accessing the `theUnsafe` private field, instead of trying to invoke `getUnsafe()`, to get an instance of `Unsafe`. Then use the instance to call `invokeCleaner` as before.

I can open a PR if the proposal makes sense, @asturio @andreasrosdal. I'd like to backport the changes (at least to the Java 17 branch) because I'm maintaining an application that's currently on Java 17.

**Note:**
`sun.misc.Unsafe#invokeCleaner` is deprecated for removal as of Java 23 and invoking it will cause a warning to be printed for Java 24+ users, but the safe alternative is only available as of Java 22: [Foreign Function & Memory API](https://openjdk.org/jeps/454). The FFM API will become an option once OpenPDF adopts a Java 22 (or newer) baseline.

---
### FFM API Deeper Dive

The `sun.misc.Unsafe#invokeCleaner` deprecation notice states,
> Use a `MemorySegment` allocated in an `Arena` with the appropriate temporal bounds.

I think a confined arena (`Arena.ofConfined()`) would be the appropriate choice in the context of this issue.
> A confined arena provides a bounded and deterministic lifetime: It is alive from the time the client opens the arena until the time the client closes the arena. A memory segment allocated in a confined arena can be accessed only before the arena is closed, at which point the region of memory backing the segment is deallocated.

A confined arena has one caveat:
> Only one thread can access the memory segments allocated in a confined arena. If multiple threads need access to a segment then a shared arena can be used. The memory segments allocated in a shared arena can be accessed by multiple threads, and any thread — whether it accesses the region or not — can close the arena to deallocate the segments.

But a shared arena doesn't guarantee immediate deallocation, according to JEP 454. (Interesting to note that the [`Arena` documentation](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/foreign/Arena.html) makes no such statement, so it's not clear if this is in fact the case.)
> Closing the [shared] arena atomically invalidates the segments, though the deallocation of the regions of memory backing the segments might not occur immediately since an expensive synchronization operation is needed to detect and cancel pending concurrent access operations on the segments.

I don't know what the thread-safety requirements/guarantees of `PdfReader`, `RandomAccessFileOrArray`, `MappedRandomAccessFile`, `LongMappedByteBuffer`, etc. are. If only a single thread will access the memory segments, then a confined arena would be suitable.
